network/ocsp.go

```diff
@@ -0,0 +1,334 @@
+		}
+	} else {
+		if len(cert.OCSPServer) > 0 {
+			log.Debugf("OCSP: Ignoring %d OCSP servers from certificate", len(cert.OCSPServer))
+		}
+	}
+
+	if v.Config.url != "" {
+		serverToCheck := ocspServerToCheck{url: v.Config.url, fromCert: false}
+
+		if v.Config.fromCert == ocspFromCertPrefer || v.Config.fromCert == ocspFromCertTrust {
+			log.Debugf("OCSP: appending server %s, from config", serverToCheck.url)
+			serversToCheck = append(serversToCheck, serverToCheck)
+		} else {
+			log.Debugf("OCSP: prepending server %s, from config", serverToCheck.url)
+			serversToCheck = append([]ocspServerToCheck{serverToCheck}, serversToCheck...)
+		}
+	}
+
+	queriedOCSPs := make(map[string]struct{})
+
+	confirms := 0
+
+	for _, serverToCheck := range serversToCheck {
+		log.Debugf("OCSP: Trying server %s", serverToCheck.url)
+
+		if _, ok := queriedOCSPs[serverToCheck.url]; ok {
+			log.Debugln("OCSP: Skipping, already queried")
+			continue
+		}
+
+		response, err := v.Client.Query(cert.Issuer.CommonName, cert, issuer, serverToCheck.url)
+		if err != nil {
+			log.WithError(err).WithField("uri", serverToCheck.url).Warnln("Cannot query OCSP server")
+
+			if v.Config.required == ocspRequiredAll {
+				return ErrOCSPRequiredAllButGotError
+			}
+
+			continue
+		}
+
+		switch response.Status {
+		case ocsp.Good:
+			confirms++
+
+			if serverToCheck.fromCert {
+				log.Debugln("OCSP: confirmed by server from certificate")
+			} else {
+				log.Debugln("OCSP: confirmed by server from config")
+			}
+
+			if v.Config.required != ocspRequiredAll {
+				// One confirmation is enough if we don't require all OCSP servers to confirm the certificate validity
+				break
+			}
+
+			if serverToCheck.fromCert && v.Config.fromCert == ocspFromCertTrust {
+				// If this OCSP server came from certificate and `--tls_ocsp_from_cert=trust`, don't perform further checks
+				break
+			}
+		case ocsp.Revoked:
+			// If any OCSP server replies with "certificate was revoked", return error immediately
+			log.WithField("serial", cert.SerialNumber).WithField("revoked_at", response.RevokedAt).Warnln("OCSP: Certificate was revoked")
+			return ErrCertWasRevoked
+		case ocsp.Unknown:
+			// Treat "Unknown" response as error if tls_ocsp_required is "yes" or "all"
+			if v.Config.required != ocspRequiredNo {
+				log.WithField("uri", serverToCheck.url).WithField("serial", cert.SerialNumber).Warnln("OCSP server doesn't know about certificate")
+				return ErrOCSPUnknownCertificate
+			}
+		}
+
+		queriedOCSPs[serverToCheck.url] = struct{}{}
+	}
+
+	if len(serversToCheck) > 0 && confirms == 0 {
+		return ErrOCSPNoConfirms
+	}
+	return nil
+}
+
+// Verify ensures certificate is not revoked by querying configured OCSP servers
+func (v DefaultOCSPVerifier) Verify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	for _, chain := range verifiedChains {
+		if len(chain) == 1 {
+			// This one cert[0] must be trusted since it was allowed by more basic verifying routines.
```

Comment:
hm, why it's trusted? it verified by expiration and signature, but where it checked for revocation status? not always clients sends own certificate with issuer. common case when only own certificate and it should be checked by CA cert which previously was distributed

Reply:
As [crypto.tls.ConnectionState](https://pkg.go.dev/crypto/tls#ConnectionState) struct description says:
> // VerifiedChains is a list of one or more chains where the first element is
> // PeerCertificates[0] and the last element is from Config.RootCAs (on the
> // client side) or Config.ClientCAs (on the server side).
> // ...
> VerifiedChains [][]*x509.Certificate

and this `VerifiedChains` stored in `tls.ConnectionState` is essentially the same as we get in this function. But since we got only one certificate in chain, that means it must be one of those `Config.RootCAs (on the client side) or Config.ClientCAs (on the server side)`, that's why I consider it "trusted"

If other side sent only one certificate, not the whole chain, TLS implementation should build whole chain to verify its path to one of root certs. If it's not possible (i.e. there exists some intermediate CA, other side didn't send its cert and we also don't have it), then we wouldn't be able to get to this point at all.

Also, [crypto.tls.Config](https://pkg.go.dev/crypto/tls#Config) says about its field `VerifyPeerCertificate`:
> // ...
> // It receives the raw ASN.1 certificates provided by the peer and also
> // any verified chains that normal processing found.
> // ...
> // If normal verification fails then the handshake will abort before considering this callback.
> // If normal verification is disabled by
> // setting InsecureSkipVerify, or (for a server) when ClientAuth is
> // RequestClientCert or RequireAnyClientCert, then this callback will
> // be considered but the verifiedChains argument will always be nil

So with `normal verification` meaning "check that leaf cert is signed by CA and CA is trusted (root) or signed by another CA (and so on)", `verifiedChains` in this function will either contain really full chains ending with ultimately trusted certificates, or be empty (which also means one doesn't care about certificate validation) and OCSP/CRL verifiers will do nothing.

I hope all this text makes at least some sense %)

Reply:
> If other side sent only one certificate, not the whole chain, TLS implementation should build whole chain to verify its path to one of root certs.

But what if the whole chain consists of just one certificate? That is, the other side is using a self-signed certificate (which is still in our trusted root store). Like, you're just using a public key and all the PKI stuff is there just for compatibility with TLS.

The fact that a certificate is in trusted root store does not mean that it's valid. Expired certificates can be placed there, it does not mean that we should trust them past their validity date (or any certificates issued outside of the validity date). Similarly, a root certificate could have been 'self-revoked', with a revocation being included into CRLs, with notices distributed to OCSPs. While this does not make much sense from the PKI perspective, where CA is supposed to revoke certificates they publish, not commit sudoku. However, it seems to be *technically* possible.

Hence, even if this a rare use case – I do realize that this should not really happen in the wild, not with proper PKI – the new code here has a branch to handle this special case. That's why I'm haggling so much about the expected behavior. If we do put something in there it has to be sensible.

Reply:
Well, I guess you're right that root CA can generate CRL with itself being in list of revoked. But I gotta check this quirky scenario to be sure (or find this in RFC).
And speaking about OCSP, is there any way CA can respond about itself? Or its dedicated OCSP-responder certificate can tell "my CA is revoked"? `(X) Doubt`
I guess it's better to dig into RFC again than think how it could work.

Reply:
CA may respond about itself and OCSP response will be signed by expired cert and we should deny such connections without success check. So it correctly maps to our current logic.